File: python/planner/weekly.py
"""주간 요약 — 한 주를 한 장으로.

금요일마다 손으로 세시던 것(이번 주 계약 몇 건, 출고 몇 건, 수수료 얼마,
다음 주에 뭐가 오는지)을 시트에서 그대로 뽑는다. 새로 입력하실 것은 없다.

브리핑과 같은 방식으로 **내용을 먼저 만들고**(`sections`) 평문과 HTML 두 가지로
그린다. 두 벌을 따로 쓰면 반드시 어긋난다 — 브리핑에서 이미 겪었다.

한 주는 **월요일 시작**이다. 주간 실적을 월~일로 세는 것이 업계 관행이고,
일요일 시작으로 두면 금요일에 보는 요약에 다음 주 이틀이 섞여 들어온다.
"""

from __future__ import annotations

from datetime import date, timedelta
from html import escape as _esc

from . import sheets, theme

WEEKDAY_KO = ["월", "화", "수", "목", "금", "토", "일"]

# 진행현황이 이 말로 시작하면 '끝난 계약' 으로 본다
DONE_WORDS = ("출고", "완료")
CANCEL_WORD = "취소"


def week_range(today: date, offset: int = 0) -> tuple[date, date]:
    """그 주의 (월요일, 일요일). offset=+1 이면 다음 주."""
    mon = today - timedelta(days=today.weekday()) + timedelta(days=7 * offset)
    return mon, mon + timedelta(days=6)


def _label(d: date) -> str:
    return f"{d:%m-%d}({WEEKDAY_KO[d.weekday()]})"


def _money(n: int) -> str:
    return f"{n:,}원"


def _fee(cr) -> int:
    """그 건의 수수료 합계 (시트 H열). 못 읽으면 0."""
    try:
        return int(sheets.digits_only(cr.total) or 0)
    except Exception:
        return 0


def _live(cr) -> bool:
    """취소된 계약이 아닌가 — 실적·예정 어디에도 취소 건은 넣지 않는다."""
    return cr.get("status").strip() != CANCEL_WORD


def _in(d, lo: date, hi: date) -> bool:
    return d is not None and lo <= d <= hi


def sections(rows: list, today: date, expiry_months: int = 3) -> list[dict]:
    """주간 요약의 내용. 브리핑 섹션과 같은 모양이라 그리는 코드를 나눠 쓴다."""
    mon, sun = week_range(today)
    nmon, nsun = week_range(today, 1)
    rows = [r for r in (rows or []) if _live(r)]

    secs = []

    # ---- 이번 주 출고 (실적) ----
    out = []
    for cr in rows:
        d = sheets.parse_date(cr.get("deliver_date"))
        if _in(d, mon, sun):
            out.append((d, cr))
    out.sort(key=lambda x: x[0])
    fee_sum = sum(_fee(cr) for _d, cr in out)
    secs.append({
        "icon": "🚗", "title": "이번 주 출고", "color": "green",
        "items": [{"lead": _label(d), "text": cr.get("customer"),
                   "sub": " · ".join(x for x in (cr.get("finance"),
                                                 cr.get("model")) if x)}
                  for d, cr in out],
        "note": f"수수료 합계 {_money(fee_sum)}" if out else "",
        "empty": "이번 주 출고가 없습니다",
    })

    # ---- 이번 주 계약(발주) ----
    made = []
    for cr in rows:
        d = sheets.parse_date(cr.get("contract_date"))
        if _in(d, mon, sun):
            made.append((d, cr))
    made.sort(key=lambda x: x[0])
    secs.append({
        "icon": "📝", "title": "이번 주 계약", "color": "blue",
        "items": [{"lead": _label(d), "text": cr.get("customer"),
                   "sub": " · ".join(x for x in (cr.get("finance"),
                                                 cr.get("model")) if x)}
                  for d, cr in made],
        "empty": "이번 주 신규 계약이 없습니다",
    })

    # ---- 금융사별 (이번 주 출고 기준) ----
    by_fin: dict[str, list] = {}
    for _d, cr in out:
        by_fin.setdefault(cr.get("finance").strip() or "기타", []).append(cr)
    if by_fin:
        order = sorted(by_fin.items(), key=lambda kv: (-len(kv[1]), kv[0]))
        secs.append({
            "icon": "🏦", "title": "금융사별", "color": "gray",
            "items": [{"lead": f"{len(v)}건", "text": k,
                       "sub": _money(sum(_fee(c) for c in v))}
                      for k, v in order],
            "count": len(order),
        })

    # ---- 다음 주 출고 예정 ----
    plan = []
    for cr in rows:
        d = sheets.parse_date(cr.get("deliver_date"))
        if _in(d, nmon, nsun):
            plan.append((d, cr))
    plan.sort(key=lambda x: x[0])
    secs.append({
        "icon": "📦", "title": "다음 주 출고 예정", "color": "orange",
        "items": [{"lead": _label(d), "text": cr.get("customer"),
                   "sub": " · ".join(x for x in (cr.get("finance"),
                                                 cr.get("model")) if x)}
                  for d, cr in plan],
        "empty": "다음 주 출고 예정이 없습니다",
    })

    # ---- 진행 중 (발주했는데 아직 출고 안 됨) ----
    open_rows = []
    for cr in rows:
        st = cr.get("status").strip()
        if any(st.startswith(w) for w in DONE_WORDS):
            continue
        cd = sheets.parse_date(cr.get("contract_date"))
        if cd is None or cd > today:
            continue
        if sheets.parse_date(cr.get("deliver_date")) is not None:
            continue                   # 출고일이 이미 잡혔으면 진행 중이 아니다
        open_rows.append(((today - cd).days, cd, cr))
    open_rows.sort(key=lambda x: -x[0])           # 오래 묵은 것이 위로
    if open_rows:
        secs.append({
            "icon": "⏳", "title": "진행 중 (출고 대기)", "color": "red",
            "items": [{"lead": f"{n}일째", "text": cr.get("customer"),
                       "sub": f"{cd:%m-%d} 계약 · {cr.get('status') or '진행현황 없음'}"}
                      for n, cd, cr in open_rows[:15]],
            "count": len(open_rows),
            "more": max(0, len(open_rows) - 15),
            "hint": "오래 묵은 건부터 캐피탈에 진행 상황을 확인해 보세요.",
        })

    # ---- 다음 주 만기 (재계약) ----
    if int(expiry_months or 0) > 0:
        exp = []
        for cr in rows:
            d = sheets.expiry_date(cr)
            if _in(d, nmon, nsun):
                exp.append((d, cr))
        exp.sort(key=lambda x: x[0])
        if exp:
            secs.append({
                "icon": "🔔", "title": "다음 주 만기", "color": "violet",
                "items": [{"lead": _label(d), "text": cr.get("customer"),
                           "sub": cr.get("model")} for d, cr in exp],
                "count": len(exp),
                "hint": "[고객관리] 탭에서 [이력] 로 상담 내역을 보고 재계약을 준비하세요.",
            })

    return secs


def title_line(today: date) -> str:
    mon, sun = week_range(today)
    return f"{mon:%Y년 %m월 %d일} ~ {sun:%m월 %d일} 주간 요약"


def to_text(rows: list, today: date, expiry_months: int = 3) -> str:
    """복사·파일 저장용 평문."""
    secs = sections(rows, today, expiry_months)
    lines = [title_line(today), ""]
    for s in secs:
        n = s.get("count", len(s["items"]))
        lines.append(f"[{s['title']}]  {n}건" if n else f"[{s['title']}]")
        if s.get("note"):
            lines.append(f"  {s['note']}")
        if not s["items"]:
            lines.append(f"  ({s.get('empty', '없음')})")
        for it in s["items"]:
            lead = (it.get("lead") or "").strip()
            sub = (it.get("sub") or "").strip()
            lines.append(f"  · {lead + '  ' if lead else ''}{it['text']}"
                         + (f"  ({sub})" if sub else ""))
        if s.get("more"):
            lines.append(f"  … 외 {s['more']}건")
        if s.get("hint"):
            lines.append(f"  → {s['hint']}")
        lines.append("")
    return "\n".join(lines).rstrip() + "\n"


def to_html(rows: list, today: date, expiry_months: int = 3) -> str:
    """화면용 — 브리핑과 같은 색·크기 규칙."""
    secs = sections(rows, today, expiry_months)
    txt, sub_c, line_c = theme.c("text"), theme.c("subtext"), theme.c("border")

    out = [f"<div style='font-family:Malgun Gothic;color:{txt};'>",
           f"<p style='margin:0 0 10px 0;font-size:19px;font-weight:bold;color:{txt};'>"
           f"{_esc(title_line(today))}</p>"]
    for s in secs:
        col, bg = theme.strong(s["color"]), theme.strong_bg(s["color"])
        n = s.get("count", len(s["items"]))
        badge = (f"<span style='background-color:{bg};color:{col};font-size:12px;"
                 f"font-weight:bold;'>&nbsp;{n}건&nbsp;</span>" if n else
                 f"<span style='color:{sub_c};font-size:12px;'>&nbsp;0건&nbsp;</span>")
        out.append(f"<hr style='border:1px solid {line_c};'>")
        out.append(f"<p style='margin:8px 0 4px 0;font-size:16px;font-weight:bold;"
                   f"color:{col};'>{s['icon']} {_esc(s['title'])} &nbsp;{badge}</p>")
        if s.get("note"):
            out.append(f"<p style='margin:0 0 4px 6px;font-size:13px;font-weight:bold;"
                       f"color:{col};'>{_esc(s['note'])}</p>")
        if not s["items"]:
            out.append(f"<p style='margin:2px 0 6px 12px;font-size:13px;color:{sub_c};'>"
                       f"{_esc(s.get('empty', '없음'))}</p>")
        for it in s["items"]:
            lead = (it.get("lead") or "").strip()
            sub = (it.get("sub") or "").strip()
            head = (f"<b style='color:{col};'>{_esc(lead)}</b> &nbsp;" if lead else "")
            tail = (f" <span style='color:{sub_c};font-size:12px;'>{_esc(sub)}</span>"
                    if sub else "")
            out.append(f"<p style='margin:1px 0 1px 12px;font-size:14px;'>"
                       f"{head}{_esc(it['text'])}{tail}</p>")
        if s.get("more"):
            out.append(f"<p style='margin:2px 0 4px 12px;font-size:12px;color:{sub_c};'>"
                       f"… 외 {s['more']}건</p>")
        if s.get("hint"):
            out.append(f"<p style='margin:4px 0 6px 12px;font-size:12px;color:{sub_c};'>"
                       f"→ {_esc(s['hint'])}</p>")
    out.append("</div>")
    return "".join(out)


def file_name(today: date) -> str:
    mon, _sun = week_range(today)
    return f"주간요약_{mon:%Y%m%d}.txt"


def week_key(today: date) -> str:
    """'이 주를 이미 보여 줬는가' 를 기억할 열쇠 (월요일 날짜)."""
    mon, _sun = week_range(today)
    return f"{mon:%Y-%m-%d}"

File: web/js/weekly.js
// 주간 요약 — planner/weekly.py 의 sections() 를 그대로 옮긴 것.
//
// 새 통신이 없다. 폰이 이미 갖고 있는 고객 줄만 가지고 계산한다.
//
// ⚠️ 파이썬과 한 건도 달라지면 안 된다. 주 경계(월요일 시작)·만기 계산·
//    취소/보류 제외 같은 규칙이 어긋나면 '이번 주 출고' 건수가 달라지는데,
//    그건 눈에 잘 안 띈다. web/verify/vectors.json 에 파이썬이 낸 정답표가 있고
//    verify_logic.mjs 가 대조한다.
import { parseDate, digitsOnly, contractMonths } from "./fmt.js";

export const WEEKDAY_KO = ["월", "화", "수", "목", "금", "토", "일"];

// 진행현황이 이 말로 시작하면 '끝난 계약'
const DONE_WORDS = ["출고", "완료"];
// 이 말로 시작하면 '내가 재촉할 건이 아니다'
const HOLD_WORDS = ["진행보류", "보류", "홀딩"];
const CANCEL_WORD = "취소";
export const EXPIRY_MONTHS = 6;

const ymd = (d) => `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}`
                 + `-${String(d.getDate()).padStart(2, "0")}`;
const md = (d) => `${String(d.getMonth() + 1).padStart(2, "0")}-`
                + `${String(d.getDate()).padStart(2, "0")}`;

/** 그 주의 [월요일, 일요일].
 *  ⚠️ 한 주는 **월요일 시작**이다. 일요일 시작으로 두면 금요일에 보는 요약에
 *     다음 주 이틀이 섞여 들어온다. */
export function weekRange(today, offset = 0) {
  // JS 의 getDay() 는 일=0 이라 파이썬 weekday()(월=0)와 다르다. 맞춰 준다.
  const wd = (today.getDay() + 6) % 7;
  const mon = new Date(today.getFullYear(), today.getMonth(),
                       today.getDate() - wd + 7 * offset);
  const sun = new Date(mon.getFullYear(), mon.getMonth(), mon.getDate() + 6);
  return [mon, sun];
}

/** d 에 months 개월을 더한다 (말일 보정). followup.add_months 와 같은 규칙. */
export function addMonths(d, months) {
  let month = d.getMonth() + months;
  const year = d.getFullYear() + Math.floor(month / 12);
  month = ((month % 12) + 12) % 12;
  const last = new Date(year, month + 1, 0).getDate();   // 그 달의 마지막 날
  return new Date(year, month, Math.min(d.getDate(), last));
}

const label = (d) => `${md(d)}(${WEEKDAY_KO[(d.getDay() + 6) % 7]})`;
const money = (n) => `${Number(n || 0).toLocaleString("ko-KR")}원`;
const get = (cr, k) => String((cr.values || {})[k] || "");

function fee(cr) {
  const n = parseInt(digitsOnly(cr.total || ""), 10);
  return isFinite(n) ? n : 0;
}

const live = (cr) => get(cr, "status").trim() !== CANCEL_WORD;

// 날짜만 견준다 — 시각이 섞이면 같은 날이 경계에서 갈린다
const between = (d, lo, hi) => d !== null && d >= lo
  && d <= new Date(hi.getFullYear(), hi.getMonth(), hi.getDate(), 23, 59, 59);

/** 계약 만기일. 계산할 수 없으면 null. (sheets.expiry_date) */
export function expiryDate(cr) {
  const n = contractMonths(get(cr, "terms"));
  if (!n) return null;
  const base = parseDate(get(cr, "deliver_date")) || parseDate(get(cr, "contract_date"));
  if (!base) return null;
  return addMonths(base, n);
}

const days = (a, b) => Math.round((a - b) / 86400000);

/** 주간 요약의 내용. weekly.sections 와 같은 모양을 돌려준다. */
export function sections(rows, today, expiryMonths = EXPIRY_MONTHS) {
  const [mon, sun] = weekRange(today);
  rows = (rows || []).filter(live);
  const secs = [];
  const who = (cr) => [get(cr, "finance"), get(cr, "model")].filter(Boolean).join(" · ");

  // ---- 이번 주 출고 ----
  const out = [];
  for (const cr of rows) {
    const d = parseDate(get(cr, "deliver_date"));
    if (between(d, mon, sun)) out.push([d, cr]);
  }
  out.sort((a, b) => a[0] - b[0]);
  const feeSum = out.reduce((s, [, cr]) => s + fee(cr), 0);
  secs.push({
    icon: "🚗", title: "이번 주 출고", color: "green",
    items: out.map(([d, cr]) => ({ lead: label(d), text: get(cr, "customer"), sub: who(cr) })),
    note: out.length ? `수수료 합계 ${money(feeSum)}` : "",
    empty: "이번 주 출고가 없습니다",
  });

  // ---- 이번 주 계약 ----
  const made = [];
  for (const cr of rows) {
    const d = parseDate(get(cr, "contract_date"));
    if (between(d, mon, sun)) made.push([d, cr]);
  }
  made.sort((a, b) => a[0] - b[0]);
  secs.push({
    icon: "📝", title: "이번 주 계약", color: "blue",
    items: made.map(([d, cr]) => ({ lead: label(d), text: get(cr, "customer"), sub: who(cr) })),
    empty: "이번 주 신규 계약이 없습니다",
  });

  // ---- 진행 중 (발주했는데 아직 출고 안 됨) ----
  const open = [];
  for (const cr of rows) {
    const st = get(cr, "status").trim();
    if (DONE_WORDS.some((w) => st.startsWith(w))) continue;
    if (HOLD_WORDS.some((w) => st.startsWith(w))) continue;   // 보류는 기다리는 건
    const cd = parseDate(get(cr, "contract_date"));
    if (cd === null || cd > today) continue;
    if (parseDate(get(cr, "deliver_date")) !== null) continue;
    open.push([days(today, cd), cd, cr]);
  }
  open.sort((a, b) => b[0] - a[0]);          // 오래 묵은 것이 위로
  if (open.length) {
    secs.push({
      icon: "⏳", title: "진행 중 (출고 대기)", color: "red",
      items: open.slice(0, 15).map(([n, cd, cr]) => ({
        lead: `${n}일째`, text: get(cr, "customer"),
        sub: `${md(cd)} 계약 · ${get(cr, "status") || "진행현황 없음"}`,
      })),
      count: open.length,
      more: Math.max(0, open.length - 15),
      hint: "오래 묵은 건부터 캐피탈에 진행 상황을 확인해 보세요.",
    });
  }

  // ---- 만기 예정 ----
  const months = Math.max(1, parseInt(expiryMonths, 10) || EXPIRY_MONTHS);
  const limit = addMonths(today, months);
  const exp = [];
  for (const cr of rows) {
    const d = expiryDate(cr);
    if (between(d, today, limit)) exp.push([d, cr]);
  }
  exp.sort((a, b) => a[0] - b[0]);
  secs.push({
    icon: "🔔", title: `${months}개월 이내 만기 예정`, color: "violet",
    items: exp.slice(0, 20).map(([d, cr]) => ({
      lead: `D-${days(d, today)}`, text: get(cr, "customer"),
      sub: [ymd(d), get(cr, "model")].filter(Boolean).join(" · "),
    })),
    count: exp.length,
    more: Math.max(0, exp.length - 20),
    empty: `${months}개월 안에 만기가 오는 고객이 없습니다`,
    // ⚠️ 파이썬과 **글자까지 같게** 둔다. 폰에 안 맞는 말(PC 의 탭 이름 등)은
    //    화면 쪽에서 바꿔 넣는다. 여기서 바꾸면 정답표 대조가 통째로 어긋나
    //    진짜 차이를 못 보게 된다 — 실제로 그렇게 5건이 한꺼번에 빨개졌다.
    hint: "[고객관리] 탭에서 [이력] 로 상담 내역을 보고 재계약을 준비하세요.",
  });

  return secs;
}

/** "2026년 09월 14일 ~ 09월 20일 주간 요약" — weekly.title_line 과 글자까지 같다. */
export function titleLine(today) {
  const [mon, sun] = weekRange(today);
  const p2 = (n) => String(n).padStart(2, "0");
  return `${mon.getFullYear()}년 ${p2(mon.getMonth() + 1)}월 ${p2(mon.getDate())}일`
       + ` ~ ${p2(sun.getMonth() + 1)}월 ${p2(sun.getDate())}일 주간 요약`;
}
